br/pkg/checksum/executor.go

```diff
@@ -0,0 +1,321 @@
+	apiver := exec.apiVersion
+	if apiver == kvrpcpb.APIVersion_V1TTL {
+		apiver = kvrpcpb.APIVersion_V1
+	}
+	resp, err := client.RawChecksum(ctx, &kvrpcpb.RawChecksumRequest{
+		Context: &kvrpcpb.Context{
+			RegionId:    regionInfo.Region.Id,
+			RegionEpoch: regionInfo.Region.RegionEpoch,
+			Peer:        peer,
+			ApiVersion:  apiver,
+		},
+		Algorithm: kvrpcpb.ChecksumAlgorithm_Crc64_Xor,
+		Ranges: []*kvrpcpb.KeyRange{{
+			StartKey: rangeStart,
+			EndKey:   rangeEnd,
+		}},
+	})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if resp.GetRegionError() != nil {
+		return errors.New(resp.GetRegionError().String())
+	}
+	if resp.GetError() != "" {
+		return errors.New(resp.GetError())
+	}
+	progressCallBack(backup.RegionUnit)
+	exec.mutex.Lock() // lock for updating checksum
+	defer exec.mutex.Unlock()
+	finalChecksum.UpdateChecksum(resp.GetChecksum(), resp.GetTotalKvs(), resp.GetTotalBytes())
+	logutil.CL(ctx).Info("region checksum finish", zap.Uint64("region id", regionInfo.Region.Id),
+		logutil.Key("region start", rangeStart),
+		logutil.Key("region end", rangeEnd),
+		zap.Reflect("checksum", finalChecksum))
+	return nil
+}
+
+const (
+	MaxScanCntLimit = uint32(1024) // limited by grpc message size
+)
+
+func (exec *Executor) doScanChecksum(
+	ctx context.Context,
+	regionInfo *restore.RegionInfo,
+	splitClient restore.SplitClient,
+	finalChecksum *Checksum,
+	progressCallBack func(backup.ProgressUnit),
+) error {
+	var peer *metapb.Peer
+	if regionInfo.Leader != nil {
+		peer = regionInfo.Leader
+	} else {
+		if len(regionInfo.Region.Peers) == 0 {
+			return errors.Annotate(berrors.ErrRestoreNoPeer, "region does not have peer")
+		}
+		peer = regionInfo.Region.Peers[0]
+	}
+	storeID := peer.GetStoreId()
+	store, err := splitClient.GetStore(ctx, storeID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	conn, err := grpc.Dial(store.GetAddress(), grpc.WithInsecure())
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer conn.Close()
+
+	client := tikvpb.NewTikvClient(conn)
+	rangeStart, rangeEnd := adjustRegionRange(exec.startKey, exec.endKey, regionInfo.Region.StartKey, regionInfo.Region.EndKey)
+	curStart := rangeStart
+	firstLoop := true
+	checksum := Checksum{}
+	digest := crc64.New(crc64.MakeTable(crc64.ECMA))
+	apiver := exec.apiVersion
+	if apiver == kvrpcpb.APIVersion_V1TTL {
+		apiver = kvrpcpb.APIVersion_V1
+	}
+	kvrpcpbCtx := &kvrpcpb.Context{
+		RegionId:    regionInfo.Region.Id,
+		RegionEpoch: regionInfo.Region.RegionEpoch,
+		Peer:        peer,
+		ApiVersion:  apiver,
+	}
+	for {
+		resp, err := client.RawScan(ctx, &kvrpcpb.RawScanRequest{
+			Context:  kvrpcpbCtx,
+			StartKey: curStart,
+			EndKey:   rangeEnd,
+			Cf:       "default",
+			Limit:    MaxScanCntLimit,
+		})
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if resp.GetRegionError() != nil {
+			return errors.New(resp.GetRegionError().String())
+		}
+		iterKvPairs := resp.GetKvs()
+		if !firstLoop { // first is duplicated
```

Comment:
Use `append(iterKvPairs[kvCnt-1].Key, 0)` as next `curStart`. Then we will not get the last key again.

Reply:
done